native_client/tflitemodelstate.cc

```diff
@@ -0,0 +1,266 @@
+    std::cerr << "Error at InterpreterBuilder for model file " << model_path << std::endl;
+    return DS_ERR_FAIL_INTERPRETER;
+  }
+
+  interpreter_->AllocateTensors();
+  interpreter_->SetNumThreads(4);
+
+  // Query all the index once
+  input_node_idx_       = get_input_tensor_by_name("input_node");
+  previous_state_c_idx_ = get_input_tensor_by_name("previous_state_c");
+  previous_state_h_idx_ = get_input_tensor_by_name("previous_state_h");
+  input_samples_idx_    = get_input_tensor_by_name("input_samples");
+  logits_idx_           = get_output_tensor_by_name("logits");
+  new_state_c_idx_      = get_output_tensor_by_name("new_state_c");
+  new_state_h_idx_      = get_output_tensor_by_name("new_state_h");
+  mfccs_idx_            = get_output_tensor_by_name("mfccs");
+
+  // When we call Interpreter::Invoke, the whole graph is executed by default,
+  // which means every time compute_mfcc is called the entire acoustic model is
+  // also executed. To workaround that problem, we walk up the dependency DAG
+  // from the mfccs output tensor to find all the relevant nodes required for
+  // feature computation, building an execution plan that runs just those nodes.
+  auto mfcc_plan = find_parent_node_ids(mfccs_idx_);
+  auto orig_plan = interpreter_->execution_plan();
+
+  // Remove MFCC nodes from original plan (all nodes) to create the acoustic model plan
+  auto erase_begin = std::remove_if(orig_plan.begin(), orig_plan.end(), [&mfcc_plan](int elem) {
+    return std::find(mfcc_plan.begin(), mfcc_plan.end(), elem) != mfcc_plan.end();
+  });
+  orig_plan.erase(erase_begin, orig_plan.end());
+
+  acoustic_exec_plan_ = std::move(orig_plan);
+  mfcc_exec_plan_ = std::move(mfcc_plan);
+
+  TfLiteIntArray* dims_input_node = interpreter_->tensor(input_node_idx_)->dims;
+
+  n_steps_ = dims_input_node->data[1];
+  n_context_ = (dims_input_node->data[2] - 1) / 2;
+  n_features_ = dims_input_node->data[3];
+  mfcc_feats_per_timestep_ = dims_input_node->data[2] * dims_input_node->data[3];
+
+  TfLiteIntArray* dims_logits = interpreter_->tensor(logits_idx_)->dims;
+  const int final_dim_size = dims_logits->data[1] - 1;
+  if (final_dim_size != alphabet_->GetSize()) {
+    std::cerr << "Error: Alphabet size does not match loaded model: alphabet "
+              << "has size " << alphabet_->GetSize()
+              << ", but model has " << final_dim_size
+              << " classes in its output. Make sure you're passing an alphabet "
+              << "file with the same size as the one used for training."
+              << std::endl;
+    return DS_ERR_INVALID_ALPHABET;
+  }
+
+  TfLiteIntArray* dims_c = interpreter_->tensor(previous_state_c_idx_)->dims;
+  TfLiteIntArray* dims_h = interpreter_->tensor(previous_state_h_idx_)->dims;
+  assert(dims_c->data[1] == dims_h->data[1]);
+  assert(state_size_ > 0);
+  state_size_ = dims_c->data[1];
+
+  return DS_ERR_OK;
+}
+
+void
+TFLiteModelState::copy_vector_to_tensor(const vector<float>& vec,
```

Comment:
Is it better than just `memcpy` ?

Reply:
Ok, from IRC, this is for zero padding mostly. Might be worth putting a comment.